refactor(oxlint/lsp): watch for `tsconfig` files when type aware is enabled

Sysix opened the pull request:
> This PR adds support for watching TypeScript configuration files (tsconfig.json) when type-aware linting is enabled in the language server. When typeAware option is set to true, the LSP will now automatically watch for changes to tsconfig.json files and trigger linter restarts when they change.

Sysix commented:
* **#16853** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16853?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16853?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=16853).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_language_server/src/linter/server_linter.rs:
```diff
@@ -384,6 +385,11 @@ impl Tool for ServerLinter {
         }
+
+        if options.type_aware {
+            watchers.push("**/tsconfig.json".to_string());
```
sysix is correct here

camc314 commented on crates/oxc_language_server/src/linter/server_linter.rs:
```diff
@@ -384,6 +385,11 @@ impl Tool for ServerLinter {
+
+        if options.type_aware {
+            watchers.push("**/tsconfig.json".to_string());
+        }
```
I do agree with this, but at the same time, this does add additional complexity. one tsconfig.jsjon could extend via too.json which wouldn't be matched.

Sysix replied:
> one tsconfig.jsjon could extend via too.json which wouldn't be matched.

Yes this is a big problem. We need to reimplement part of tsconfig resolver. Updated the code to watch for `**/tsconfig*.json`

camc314 commented:
### Merge activity

* **Dec 14, 9:24 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Dec 14, 9:24 PM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

Sysix committed:
refactor(oxlint/lsp): watch for `tsconfig` files when type aware is enabled (#16853)

> This PR adds support for watching TypeScript configuration files (tsconfig.json) when type-aware linting is enabled in the language server. When typeAware option is set to true, the LSP will now automatically watch for changes to tsconfig.json files and trigger linter restarts when they change.
